### PR TITLE
Reduce single file format overhead by lazily importing modules and loading .gitignore / --exclude

### LINKED ISSUE #2564: Black slow startup to format a single file
I'm a bit annoyed that it takes too much time to format a single file when using black to auto-save contents in my IDE (in this case it's Eclipse/PyDev, but I guess it should be the same for any place using the command line to format a single file).

-- In Eclipse/PyDev, the integration (I've done) is done so that for each new file to be formatted, black is executed and contents are piped in/out.

For doing this execution (in my machine), black takes `0.3s`, but out of that time, `0.255s` (`83%` of the time) is taken on doing imports, `0.025s` (`9%`) getting git info and only `0.020s` (`8%`) on actually doing the formatting...

In use case where a single file is passed is that it'll call `black.files.get_gitignore` even when a single `.py` is passed, which is pretty slow and it shouldn't be needed in this case (without it, the formatting time of `0.045s` goes to `0.02s`... this can be seen passing some dummy `--exclude` which has the side-effect of not loading the git ignore contents).

Some profiling numbers on doing a single run for a file in `black` from the command line:

image

So, some things I considered to improve that:

1. Move some of the bigger imports from top-level to inside methods so that it happens lazily (that way things unneeded such as the `asyncio` wouldn't be executed at all). Note: this would need to move some top-level imports to a `typing.TYPE_CHECKING` check too.
2. Make the `get_gitignore` called only if needed.

I guess another option could be using the black server in the integration, but given that it's not installed by default in user environments this is a bit of a hassle for end users (since it's not really polite to automatically mess user environments installing `black[d]` one needs to ask them to do so -- and it doesn't fix the issue where it's used in the command line).

So, I was considering providing a pull request for that -- to make some imports local / don't call `get_gitignore` unless needed -- is this something that seems ok? (for me it's a no-brainer, but I've seen projects adamant on using top-level imports even if it works to their detriment, so, I'd rather ask first than do the work and get rejected later).

### PR DESCRIPTION
**Pull request by ichard26**
### Description

```
`black.reformat_many` depends on a lot of slow-to-import modules. When
formatting simply a single file, the time paid to import those modules
is totally wasted. So I moved `black.reformat_many` and its helpers
to `black.concurrency` which is now *only* imported if there's more
than one file to reformat. This way, running Black over a single file
is snappier

Here are the numbers before and after this patch running `python -m
black --version`:

- interpreted: 411 ms +- 9 ms -> 342 ms +- 7 ms: 1.20x faster
- compiled: 365 ms +- 15 ms -> 304 ms +- 7 ms: 1.20x faster
```

```
Loading .gitignore and compiling the exclude regex can take more than
15ms. We shouldn't and don't need to pay this cost if we're simply
formatting files given on the command line directly.

I would've loved to lazily import pathspec, but the patch won't be clean
until the file collection and discovery logic is refactored first.

Co-authored-by: Fabio Zadrozny <fabiofz@gmail.com>
```

Fixes #2564 and closes #2656.

### Checklist - did you ...

- [x] Add a CHANGELOG entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation? -> n/a

### CODE REVIEW

**Comment by ichard26** [comment]
For what it is worth, I'd prefer a rebase merge for this PR given the commits are mostly independent (other than the changelog entry).

**Comment by ichard26** [comment]
Coolio, diff-shades is failing and I don't know why at all ...

**Comment by ichard26** [comment]
Coolio, diff-shades is failing and I don't know why at all ...

**Review by Doyhenc99 — CHANGES_REQUESTED** [review]
Sadang bib

File: src/black/__init__.py
Comment (JelleZijlstra) [inline src/black/__init__.py:94]: Hm this does a syscall at startup, we may want to put it into a cached-forever function instead.

File: src/black/__init__.py
Comment (JelleZijlstra) [inline src/black/__init__.py:1373]: Instead of this global, what if we return early in `maybe_install_uvloop` if uvloop is already in sys.modules?

Or we could just call it unconditionally. It's only called once per formatting run and it's pretty fast:

```
In [14]: %timeit black.concurrency.maybe_install_uvloop()
1.67 µs ± 30.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

File: src/black/__init__.py
Comment (JelleZijlstra) [reply src/black/__init__.py:94]: Though then again

```
In [16]: %timeit os.cpu_count()
1.05 µs ± 10.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

**Comment by ichard26** [comment]
PTAL. Please note I'd prefer a **rebase merge** here if you approve and chose to merge right after.

File: src/black/concurrency.py
Comment (ichard26) [inline src/black/concurrency.py:88]: I chose to keep the helper function and instead of always attempting an import of uvloop in the module scope just as insurance if we ever accidentally break the invariant `black.concurrency` should only be imported via `black.main` if there's more than one file to format.

**Comment by ichard26** [comment]
This works on Windows with PyInstaller (just tested locally) so I'm merging.

**Pull request merged**
